feat: Use URN for all extension URIs (#852)
substrait-io/substrait: site/docs/extensions/index.md

A YAML file can also reference types and type variations defined in another YAML file. To do this, it must declare the YAML file it depends on using a key-value pair in the `dependencies` key, where the value is the URI to the YAML file, and the key is a valid identifier that can then be used as an identifier-safe alias for the URI. This alias can then be used as a `.`-separated namespace prefix wherever a type class or type variation name is expected.

> For example, if a YAML file needs to reference a type from Substrait's default extensions (which use URNs as their preferred URI format, see [Canonical URIs](#canonical-uris-for-default-extensions) below), it would look like this. If the YAML file `extension_types.yaml` with URI `urn:substrait:extension_types` defines a type called `point`, a different YAML file can use the type in a function declaration as follows:

Reviewer: ```suggestion
For example, if a YAML file needs to reference a type from Substrait's default extensions (which use URNs as their preferred URI format, see [Canonical URIs](#canonical-uris-for-default-extensions) below). If we have a YAML file `extension_types.yaml` with URI `urn:substrait:extension_types` defines a type called `point`, a different YAML file can use the type in a function declaration as follows:
```

Reviewer: Yes, that looks good to me!

Here, the choice for the name `ext` is arbitrary, as long as it does not conflict with anything else in the YAML file.

### Canonical URIs for Default Extensions

> Substrait provides a set of default simple extensions included in the repository. These extensions use URNs (Uniform Resource Names) as their canonical URIs following the pattern `urn:substrait:<filename_without_extension>`. While downstream users can use any URI they prefer to identify these extensions, library authors **must** use the canonical URIs to ensure compatibility between Substrait plans produced and consumed across different libraries.

Reviewer 2: To be slightly future proof, perhaps add *relative path from extension directory* or something along that line? This does not change whatever described below, but leaves the door open to naturally incorporate  subdirectories when we introduce.

But as-is is also fine and we can extend this when we actually have the cases.

> Substrait provides a set of default simple extensions included in the repository. These extensions use URNs (Uniform Resource Names) as their canonical URIs following the pattern `urn:substrait:<filename_without_extension>`. While downstream users can use any URI they prefer to identify these extensions, library authors **must** use the canonical URIs to ensure compatibility between Substrait plans produced and consumed across different libraries.

Author: I added a link for the first mention of `URI` as well just for good measure 🙂
